disnake/ext/commands/interaction_bot_base.py

```diff
@@ -436,50 +600,132 @@ def get_slash_command(
             raise TypeError(f"Expected name to be str, not {name.__class__}")
 
         chain = name.split()
-        slash = self.all_slash_commands.get(chain[0])
-        if slash is None:
-            return None
-
-        if len(chain) == 1:
-            return slash
-        elif len(chain) == 2:
-            return slash.children.get(chain[1])
-        elif len(chain) == 3:
-            group = slash.children.get(chain[1])
-            if isinstance(group, SubCommandGroup):
-                return group.children.get(chain[2])
-
-    def get_user_command(self, name: str) -> Optional[InvokableUserCommand]:
+        if guild_id is not MISSING:
+            cmd_index = AppCmdIndex(
+                type=ApplicationCommandType.chat_input, name=chain[0], guild_id=guild_id
+            )
+            command = self.all_app_commands.get(cmd_index)
+            if command is None:
+                return None
+            return _match_subcommand_chain(command, chain)  # type: ignore
+
+        # this is mostly for backwards compatibility, as previously guild_id arg didn't exist
+        result = None
+        for command in self.all_app_commands.values():
```

Comment:
Do you mean modifying the `Dict[(type, name, guild_id), Command]` to be a `Dict[(type, name), Dict[guild_id, Command]]`?

Reply:
That could be a solution, yeah - this way we could check `len(commands[(type, name)])` to see if there's more than one candidate, which means that while it'll always have to do 2 lookups, it avoids the `O(n)` case here (I think?).

I sort of want to test the memory footprint of this first though, since it would mean one dict per command in most situations

Reply:
It's still `O(n)` if there's a single command with `n` guild IDs

Reply:
The only thing that comes to mind is a `Dict[(type, name), Set[int]]`, where `Set[int]` is a set of `id(app_command)`, coexisting with our `Dict[(type, name, guild_id)]`. This way we could check `len(commands[(type, name)])`.

Reply:
mmm, true :/
For code using currently released versions there would really only be one entry each, so that structure would help in those common cases; but you're right, the worst case actually stays the same

Reply:
I agree that linear search doesn't seem like an optimal solution, but this search is done over a relatively small set of objects and it's trivial to eliminate by specifying the guild ID. In conclusion, it's a reasonable solution for backwards compatibility.